pkg/controller/master/node/promote_check.go

```diff
@@ -0,0 +1,146 @@
+package node
+
+import (
+	"sort"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type SortByAgeNodeList []*corev1.Node
+
+func (s SortByAgeNodeList) Len() int      { return len(s) }
+func (s SortByAgeNodeList) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s SortByAgeNodeList) Less(i, j int) bool {
+	// reverse order
+	return !s[i].CreationTimestamp.Before(&s[j].CreationTimestamp)
+}
+
+// selectPromoteNodeList return a list contain nodes need to be promoted
+// If the cluster doesn't need to be promoted, return nil
+func (h *Handler) selectPromoteNodeList(nodeList []*corev1.Node) ([]*corev1.Node, error) {
+	masterNumber := 0
+	ignoreNodeNumber := 0
+	canBePromoteNodeList := make([]*corev1.Node, 0, len(nodeList))
+	for _, node := range nodeList {
+		if h.shouldBeMaster(node) {
+			masterNumber++
+		} else {
+			if h.isHarvesterNode(node) {
```

Comment:
Check earlier before listing all nodes

Reply:
This logic is take care about user already has some master (not installed by harvester-installer)